### PR TITLE
[DataPipe] Fixing pickling issues with fork and demux

### LINKED ISSUE #67848: `Generator` needs to be removed after iteration ends for `Demux` and `Forker`
## 🐛 Bug

The `generator` attached to `Demux` or `Forker` is never cleaned up.
https://github.com/pytorch/pytorch/blob/8e1ead8e4df3fa8acddd13c8b68d5c26690a9ec1/torch/utils/data/datapipes/iter/combining.py#L108-L110
https://github.com/pytorch/pytorch/blob/8e1ead8e4df3fa8acddd13c8b68d5c26690a9ec1/torch/utils/data/datapipes/iter/combining.py#L256-L258

The generator object is attached to `DataPipe` instance. If we serialize or `traverse` the datapipe graph before iterating the pipeline, there won't be any error. But, if we run through an iteration, the `generator` object is initialized but not cleaned up, it would complain the `generator` object can not be pickled.

To reproduce:
```py
>>> from torch.utils.data.datapipes.iter import IterableWrapper
>>> from torch.utils.data.datapipes.graph import traverse

>>> dp = IterableWrapoer([1,2,3])
>>> dp1, dp2 = dp.fork(2)

>>> print(traverse(dp1))  # Still working

>>> for _ in zip(dp1, dp2):
...     pass

>>> print(traverse(dp1))
TypeError: cannot pickle 'generator' object
```

cc @VitalyFedyunin @ejguan @NivekT

### PR DESCRIPTION
**Pull request by NivekT**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67930

Fixes #67848

Differential Revision: [D32222184](https://our.internmc.facebook.com/intern/diff/D32222184)

cc @VitalyFedyunin @ejguan @NivekT

### CODE REVIEW

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32222184).

File: test/test_datapipe.py [hunk 1]
```diff
@@ -862,6 +864,11 @@ def test_fork_datapipe(self):
         self.assertEqual(len(input_dp), len(dp2))
         self.assertEqual(len(input_dp), len(dp3))
 
+        # Pickle Test:
+        dp1, dp2, dp3 = input_dp.fork(num_instances=3)
+        from torch.utils.data.graph import traverse
+        traverse(dp1)  # This should not raise any error
```
Comment (ejguan) [inline test/test_datapipe.py:870]: Could we add another test that we iterates `dp1`, `dp2` and `dp3` and then traverse?

**Review by ejguan — APPROVED** [review]
LGTM, thanks for fixing the issue.
I only have one comment on test case.

File: test/test_datapipe.py (hunk 1 above)
Comment (NivekT) [reply test/test_datapipe.py:870]: Done

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32222184).

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32222184).

File: test/test_datapipe.py [hunk 2]
```diff
@@ -52,6 +52,7 @@
     runtime_validation,
     runtime_validation_disabled,
 )
+from torch.utils.data.graph import traverse
```
Comment (ejguan) [inline test/test_datapipe.py:55]: Side note: Do you think we can also add this to one of the standard tests for each DataPipe in the future?
Maybe open an issue?

Comment (NivekT) [reply test/test_datapipe.py:55]: I think it should go here:
https://github.com/pytorch/pytorch/wiki/DataPipes-testing-requirements

One of these two?
- Test if DataPipes creates properly linked DataPipes graph
- Test if DataPipe is picklable (using pickle or dill)

Comment (ejguan) [reply test/test_datapipe.py:55]: I mean we need to test all existing and in-development DataPipes for this case.

Comment (ejguan) [reply test/test_datapipe.py:55]: After iteration, it's still picklable.

Comment (NivekT) [reply test/test_datapipe.py:55]: I see, I will add it to the testing requirement page. 

Perhaps we should have an issue to ensure all DataPipes have tests that meet all the requirements on that page? Or do you prefer to have a specific issue about being picklable after iteration for now?
I can do either.

Comment (ejguan) [reply test/test_datapipe.py:55]: I would go with former one and categorize which test is missing for each DataPipe

**Pull request closed without merging**